File: java/api/src/main/java/eu/ill/cameo/api/base/ConnectionChecker.java
/*
 * CAMEO
 *
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under BSD 3-Clause and GPL-v3 as described in license files.
 * You may not use this work except in compliance with the Licences.
 *
 */

package eu.ill.cameo.api.base;

/**
 * Class providing a simple connection checker.
 */
public class ConnectionChecker {

	public static interface Handler {
		
		void handle(boolean available);
	}
	
	private Server server;
	private Handler handler;
	private Thread thread = null;
	private TimeCondition waitCondition = new TimeCondition();
	
	ConnectionChecker(Server server, Handler handler) {
		this.server = server;
		this.handler = handler;
	}

	void start(final int timeoutMs, final int pollingTimeMs) {

		thread = new Thread(new Runnable() {
			public void run() {
	
				while (true) {
					
					boolean stopped = waitCondition.waitFor(pollingTimeMs);
					if (stopped) {
						return;
					}

					// Check the server.
					boolean available = (server.isAvailable(timeoutMs));
					
					// Apply the handler.
					handler.handle(available);
				}
			}
		});
		
		thread.start();
	}

	/**
	 * Terminates the checker.
	 */
	public synchronized void terminate() {
		if (thread != null) {
			waitCondition.notifyCondition();
			try {
				thread.join();
			}
			catch (InterruptedException e) {
			}
		}
	}
}

File: cpp/api/src/base/ConnectionChecker.cpp
/*
 * CAMEO
 *
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under BSD 3-Clause and GPL-v3 as described in license files.
 * You may not use this work except in compliance with the Licences.
 *
 */

#include "ConnectionChecker.h"

#include "Server.h"
#include "TimeCondition.h"

namespace cameo {

ConnectionChecker::ConnectionChecker(Server * server, ConnectionChecker::FunctionType handler) : m_server(server), m_function(handler) {
	m_waitCondition.reset(new TimeCondition{});
}

ConnectionChecker::~ConnectionChecker() {
	stopThread();
}

void ConnectionChecker::loop(int timeoutMs, int pollingTimeMs) {

	// Loop until the condition is notified.
	while (true) {
		bool stopped {m_waitCondition->wait(pollingTimeMs)};
		if (stopped) {
			return;
		}

		// Check the server.
		bool available {m_server->isAvailable(timeoutMs)};

		// Apply the handler.
		m_function(available);
	}
}

void ConnectionChecker::startThread(int timeoutMs, int pollingTimeMs) {

	// Stop the thread if it exists.
	stopThread();

	// Start the thread.
	m_thread = std::unique_ptr<std::thread>{new std::thread(std::bind(&ConnectionChecker::loop, this, timeoutMs, pollingTimeMs))};
}

void ConnectionChecker::stopThread() {

	if (m_thread) {
		m_waitCondition->notify();
		m_thread->join();
	}
}

}